src/ray/common/asio/io_context.h

```diff
@@ -0,0 +1,108 @@
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <boost/asio.hpp>
+#include <limits>
+#include "absl/container/flat_hash_map.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/ray_config.h"
+#include "ray/util/logging.h"
+
+/// Count, queueing, and execution statistics for an asio handler.
+struct HandlerStats {
+  // Counts.
+  int64_t cum_count = 0;
+  int64_t curr_count = 0;
+
+  // Queue stats.
+  int64_t cum_queue_time = 0;
+  int64_t min_queue_time = std::numeric_limits<int64_t>::max();
+  int64_t max_queue_time = -1;
+
+  // Execution stats.
+  int64_t cum_execution_time = 0;
+  int64_t min_execution_time = std::numeric_limits<int64_t>::max();
+  int64_t max_execution_time = -1;
+};
+
+/// A mutex wrapper around a handler stats struct.
+struct GuardedHandlerStats {
+  // Stats for some handler.
+  HandlerStats stats;
+
+  // The mutex protecting the reading and writing of these stats.
+  // This mutex should be acquired with a reader lock before reading, and should be
+  // acquired with a writer lock before writing.
+  mutable absl::Mutex mutex;
+};
+
+/// A proxy for boost::asio::io_context that collects statistics about posted handlers.
+class io_context_proxy : public boost::asio::io_context {
+ public:
+  /// A proxy post function that collects count, queueing, and execution statistics for
+  /// the given handler.
+  ///
+  /// \param handler The handler to be posted to the event loop.
+  /// \param name A human-readable name for the handler, to be used for viewing stats
+  /// for the provided handler. Defaults to UNKNOWN.
+  void post(std::function<void()> handler, const std::string &name = "UNKNOWN")
+      LOCKS_EXCLUDED(mutex_);
+
+  /// Returns a snapshot view of the global count, queueing, and execution statistics
+  /// across all handlers.
+  ///
+  /// \return A snapshot view of the global handler stats.
+  HandlerStats get_global_stats() const;
+
+  /// Returns a snapshot view of the count, queueing, and execution statistics for the
+  /// provided handler.
+  ///
+  /// \param handler_name The name of the handler whose stats should be returned.
+  /// \return A snapshot view of the handler's stats.
+  absl::optional<HandlerStats> get_handler_stats(const std::string &handler_name) const
+      LOCKS_EXCLUDED(mutex_);
+
+  /// Returns snapshot views of the count, queueing, and execution statistics for all
+  /// handlers.
+  ///
+  /// \return A vector containing snapshot views of stats for all handlers.
+  std::vector<std::pair<std::string, HandlerStats>> get_handler_stats() const
+      LOCKS_EXCLUDED(mutex_);
```

Comment:
These aren't yet used outside of the `StatsString` implementation, but seemed like they would be useful to expose publicly in case some consumers want to poke at the handler stats outside of the full stats string, e.g. if some consumer detects heavy load, it might only be interested in a subset of the event loop stats.